dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphUtility/BarGraph.cs

```diff
@@ -0,0 +1,479 @@
+        private TextFormat textFormatTitle;
+        private TextFormat textFormatHorizontal;
+        private TextFormat textFormatVertical;
+
+        private ShapeVisual shapeContainer;
+        private CompositionLineGeometry xAxisLine;
+        private CompositionLineGeometry yAxisLine;
+        private ContainerVisual mainContainer;
+
+        private static SharpDX.Mathematics.Interop.RawColor4 black = new SharpDX.Mathematics.Interop.RawColor4(0, 0, 0, 255);
+        private static SharpDX.Mathematics.Interop.RawColor4 white = new SharpDX.Mathematics.Interop.RawColor4(255, 255, 255, 255);
+
+        private static float textSize = 20.0f;
+
+        private AmbientLight ambientLight;
+        private SpotLight barOutlineLight;
+        private PointLight barLight;
+
+        #region public setters
+        public string Title { get; set; }
+        public string XAxisLabel { get; set; }
+        public string YAxisLabel { get; set; }
+        public ContainerVisual BarRoot { get; }
+        public ContainerVisual GraphRoot { get; }
+        #endregion
+
+        public enum GraphBarStyle
+        {
+            Single = 0,
+            Random = 1,
+            PerBarLinearGradient = 3,
+            AmbientAnimatingPerBarLinearGradient = 4
+        }
+
+        // Constructor for bar graph.
+        // To insert graph, call the constructor then use barGraph.Root to get the container to parent.
+        public BarGraph(Compositor compositor, IntPtr hwnd, string title, string xAxisLabel, 
+            string yAxisLabel, float width, float height, double dpiX, double dpiY, float[] data, WindowRenderTarget renderTarget, 
+            bool AnimationsOn = true, GraphBarStyle graphBarStyle = GraphBarStyle.Single,
+            List<Windows.UI.Color> barColors = null)
+        {
+            this.compositor = compositor;
+            this.hwnd = hwnd;
+            this.graphWidth = (float)(width * dpiX / 96.0);
+            this.graphHeight = (float)(height * dpiY / 96.0);
+
+            this.graphData = data;
+
+            Title = title;
+            XAxisLabel = xAxisLabel;
+            YAxisLabel = yAxisLabel;
+
+            this.graphBarStyle = graphBarStyle;
+
+            if (barColors != null)
+            {
+                graphBarColors = barColors;
+            }
+            else
+            {
+                graphBarColors = new List<Windows.UI.Color>() { Colors.Blue };
+            }
+
+            // Configure options for text.
+            var factory2D = new SharpDX.Direct2D1.Factory();
+
+            var properties = new HwndRenderTargetProperties();
+            properties.Hwnd = this.hwnd;
+            properties.PixelSize = new SharpDX.Size2((int)(width * dpiX / 96.0), (int)(width * dpiY / 96.0));
+            properties.PresentOptions = PresentOptions.None;
+
+            textRenderTarget = renderTarget;
+
+            // Generate graph structure.
+            var graphRoot = GenerateGraphStructure();
```

Comment:
Why bother with the local if you only use it once?

Reply:
condensed